README.md
![gtk logo][logo]

# Gnome::Gtk3::Glade - Accessing GTK+ using Glade
[![License](http://martimm.github.io/label/License-label.svg)](http://www.perlfoundation.org/artistic_license_2_0)

# Description
With the modules from package `Gnome::Gtk3` you can build a user interface and interact with it. This package however, is meant to load a user interface description saved by an external designer program. The program used is **glade** which saves an **XML** description of the made design.

The user must provide one or more classes containing methods to receive signals defined in the user interface design. Registration of signals will be done automatically.

Then only two lines of code (besides the loading of modules) is needed to let the user interface appear and enter the main loop.

# Synopsis
### User interface file
The first thing to do is designing a ui and save it. A part of the saved result is shown below. It shows the part of an exit button. Assume that this file is saved in **example.glade**.
```
<?xml version="1.0" encoding="UTF-8"?>
<!-- Generated with glade 3.20.0 -->
<interface>
  <requires lib="gtk+" version="3.0"/>
  <object class="GtkWindow" id="window">
...
          <object class="GtkButton" id="quit">
            <property name="label">Quit</property>
            <property name="visible">True</property>
            <property name="can_focus">False</property>
            <property name="receives_default">False</property>
            <signal name="clicked" handler="quit-program"/>
          </object>
...
</interface>

```

### Class for signal handlers
Then write code to handle all signals which are defined by the user interface. These modules are called engines. You do not have to write every handler at once. You will be notified about a missing handler as soon as an event is fired for it.

Only the method to handle a click event from the quit button is shown below in the example. This example file is saved in **lib/MyEngine.pm6**.

```
use v6;
use Gnome::Gtk3::Glade;
use Gnome::Gtk3::Glade::Engine;

unit class MyEngine;
also is Gnome::Gtk3::Glade::Engine;

# $widget is the activated button after which this method is called. Methods
# are from Gnome::Gtk3::Button. See documentation in the Gnome::Gtk3 project.
method quit-program ( :$widget ) {

  note "Button label: ", $widget.get-label;
  note "Button name is by default button's class name: ", $widget.get-name;

  self.glade-main-quit();
}
  ...
```

### The main program
The rest is a piece of cake.
```
use v6;
use MyEngine;
use Gnome::Gtk3::Glade;

my Gnome::Gtk3::Glade $gui .= new;
$gui.add-gui-file("example.glade");
$gui.add-engine(MyEngine.new);
$gui.run;
```

# Documentation

* [Gnome::Gtk3::Glade][Gnome::Gtk3::Glade pdf]
* Gnome::Gtk3::Glade::Engine

## Miscellaneous
* [Release notes](https://github.com/MARTIMM/gnome-glade3/blob/master/doc/CHANGES.md)

# TODO

* [ ] What can we do with the Gnome::Gtk3::Glade object after it exits the main loop.
* [ ] Documentation.

# Versions of involved software

* Program is tested against the latest version of **Rakudo** build on **moarvm**.
* Used **glade** version is **>= 3.22**

# Installation of Gnome::Gtk3::Glade

`zef install Gnome::Gtk3::Glade`


# Author

Name: **Marcel Timmerman**
Github account name: **MARTIMM**


<!---- [refs] ----------------------------------------------------------------->
[release]: https://github.com/MARTIMM/gnome-glade3/blob/master/doc/CHANGES.md
[logo]: https://github.com/MARTIMM/gnome-glade3/blob/master/doc/design-docs/gtk-logo-100.png

[Gnome::Gtk3::Glade pdf]: https://nbviewer.jupyter.org/github/MARTIMM/gnome-glade3/blob/master/doc/Glade3.pdf


CHANGES.md
## Release notes

* 2020-02-25 0.8.9:
  * Changes in some routines to cope with deprecated and removed parts from Gnome::Gtk3.

* 2020-01-10 0.8.8.1:
  * Repo renaming. Perl6 to Raku.

* 2019-10-16 0.8.8
  * Changes in signal and text processing caused by changes in `Gnome::*` packages.

* 2019-08-23 0.8.7
  * Bug fixes:
    * A dependency on removed deprecated code.
    * Use of an enum from the wrong gtk3 module.
    * Initialization of TextIter now with :empty.

* 2019-07-22 0.8.6
  * Bugs removed from META config. Also small bugs found in Test module and made some improvements.
  * Added a load test which creates an empty window and destroys it.

* 2019-06-01 0.8.5
  * Renamed project from Gnome::Gtk3::Glade into Gnome::Gtk3::Glade

* 2019-04-26 0.8.4
  * Small changes from Gnome::Gtk3

* 2019-04-22 0.8.3
  * Extended test xt/200-gui-read.t to use device events. This can only be tested by hand because I do not know how to create mouse button presses or generate keyboard input.

* 2019-03-11 0.8.2
  * Improve test facility
* 2019-03-10 0.8.1
  * Modifications from Gnome::Gtk3
* 2019-03-06 0.8.0
  * Changing glade interface to handle more engines, css and theme modules.
  * Added documentation for Gnome::Gtk3::Glade
* 2019-02-23 0.7.3
  * Removed convenience method `glade-get-widget()` because Gnome::Gtk3 has :build-id in the widgets new() method which does the same work.
* 2019-02-18 0.7.2
  * Bugfixes
* 2019-02-11 0.7.1
  * Conversions to use Gnome::Gtk3
* 2019-01-24 0.7.0
  * Added a Image module to the native call interface.
* 2019-01-22 0.6.3
  * Factoring out NativeGtk. It is getting too big.
* 2019-01-21 0.6.2
  * Bug fixes in the testing framework. Some are captured but not understood.
* 2019-01-19 0.6.1
  * The testing framework is improved. There are however problems using threads.
* 2019-01-18 0.6.0
  * Added basics of a testing framework. It is possible to emit a signal to some particular widget.
* 2019-01-17 0.5.0
  * Added convenience methods to `Engine` to handle text easier in te callback methods.
* 2019-01-16 0.4.0
  * Added the use of css stylesheets. It will by default get the `GTK_STYLE_PROVIDER_PRIORITY_USER` priority to override everything. There are however classes on widgets which prevent settings of e.g. background-color. That is only possible when the `flat` class is set on the widget. Each widget has its own set of classes which must be looked up from the documents. The `GtkButton`, for example, recognizes `circular` and `flat` amongst others. There is not a way to trap errors when parsing of the css fails.
* 2019-01-15 0.3.3
  * The hash of objects is not provided anymore to the callback methods. A convenience method `glade-get-widget( Str $id --> GtkWidget )` is provided in the `Gnome::Gtk3::Glade::Engine` class.
* 2019-01-12 0.3.2
  * Bugfixes
* 2019-01-07 0.3.1
  * Bugfixes
  * Added GError structure to get error messages
* 0.3.0
  * Use native call interface to display glade designed ui using GtkBuilder.
  * Get GTK widget objects from GtkBuilder using their id found from the glade description.
  * Find signal descriptions and activate them.
* 0.2.0 Read ui description and find objects and signals
* 0.1.0 Make tests written in C to study how to show a glade saved ui interface description.
* 0.0.1 Start of project


xt/cc/enum-size.c
#include <stdio.h>
#include <stdlib.h>

typedef enum
{
  GDK_SETTING_ACTION_NEW,
  GDK_SETTING_ACTION_CHANGED,
  GDK_SETTING_ACTION_DELETED
} GdkSettingAction;

typedef enum
{
  GDK_WINDOW_STATE_WITHDRAWN        = 1 << 0,
  GDK_WINDOW_STATE_ICONIFIED        = 1 << 1,
  GDK_WINDOW_STATE_MAXIMIZED        = 1 << 2,
  GDK_WINDOW_STATE_STICKY           = 1 << 3,
  GDK_WINDOW_STATE_FULLSCREEN       = 1 << 4,
  GDK_WINDOW_STATE_ABOVE            = 1 << 5,
  GDK_WINDOW_STATE_BELOW            = 1 << 6,
  GDK_WINDOW_STATE_FOCUSED          = 1 << 7,
  GDK_WINDOW_STATE_TILED            = 1 << 8,
  GDK_WINDOW_STATE_TOP_TILED        = 1 << 9,
  GDK_WINDOW_STATE_TOP_RESIZABLE    = 1 << 10,
  GDK_WINDOW_STATE_RIGHT_TILED      = 1 << 11,
  GDK_WINDOW_STATE_RIGHT_RESIZABLE  = 1 << 12,
  GDK_WINDOW_STATE_BOTTOM_TILED     = 1 << 13,
  GDK_WINDOW_STATE_BOTTOM_RESIZABLE = 1 << 14,
  GDK_WINDOW_STATE_LEFT_TILED       = 1 << 15,
  GDK_WINDOW_STATE_LEFT_RESIZABLE   = 1 << 16
} GdkWindowState;

typedef enum
{
  GDK_NOTHING		= -1,
  GDK_DELETE		= 0,
  GDK_DESTROY		= 1,
  GDK_EXPOSE		= 2,
  GDK_MOTION_NOTIFY	= 3,
  GDK_BUTTON_PRESS	= 4,
  GDK_2BUTTON_PRESS	= 5,
  GDK_DOUBLE_BUTTON_PRESS = GDK_2BUTTON_PRESS,
  GDK_3BUTTON_PRESS	= 6,
  GDK_TRIPLE_BUTTON_PRESS = GDK_3BUTTON_PRESS,
  GDK_BUTTON_RELEASE	= 7,
  GDK_KEY_PRESS		= 8,
  GDK_KEY_RELEASE	= 9,
  GDK_ENTER_NOTIFY	= 10,
  GDK_LEAVE_NOTIFY	= 11,
  GDK_FOCUS_CHANGE	= 12,
  GDK_CONFIGURE		= 13,
  GDK_MAP		= 14,
  GDK_UNMAP		= 15,
  GDK_PROPERTY_NOTIFY	= 16,
  GDK_SELECTION_CLEAR	= 17,
  GDK_SELECTION_REQUEST = 18,
  GDK_SELECTION_NOTIFY	= 19,
  GDK_PROXIMITY_IN	= 20,
  GDK_PROXIMITY_OUT	= 21,
  GDK_DRAG_ENTER        = 22,
  GDK_DRAG_LEAVE        = 23,
  GDK_DRAG_MOTION       = 24,
  GDK_DRAG_STATUS       = 25,
  GDK_DROP_START        = 26,
  GDK_DROP_FINISHED     = 27,
  GDK_CLIENT_EVENT	= 28,
  GDK_VISIBILITY_NOTIFY = 29,
  GDK_SCROLL            = 31,
  GDK_WINDOW_STATE      = 32,
  GDK_SETTING           = 33,
  GDK_OWNER_CHANGE      = 34,
  GDK_GRAB_BROKEN       = 35,
  GDK_DAMAGE            = 36,
  GDK_TOUCH_BEGIN       = 37,
  GDK_TOUCH_UPDATE      = 38,
  GDK_TOUCH_END         = 39,
  GDK_TOUCH_CANCEL      = 40,
  GDK_TOUCHPAD_SWIPE    = 41,
  GDK_TOUCHPAD_PINCH    = 42,
  GDK_PAD_BUTTON_PRESS  = 43,
  GDK_PAD_BUTTON_RELEASE = 44,
  GDK_PAD_RING          = 45,
  GDK_PAD_STRIP         = 46,
  GDK_PAD_GROUP_MODE    = 47,
  GDK_EVENT_LAST        /* helper variable for decls */
} GdkEventType;

int main ( ) {
  printf( "enum size GdkSettingAction: %d\n", sizeof(GdkSettingAction));
  printf( "enum size GdkWindowState: %d\n", sizeof(GdkWindowState));
  printf( "enum size GdkEventType: %d\n", sizeof(GdkEventType));

  exit(0);
}
